src/v/cluster/partition_properties_stm.cc

```diff
@@ -0,0 +1,326 @@
+      _log.trace,
+      "taking snapshot at offset: {}, state snapshots: {}",
+      o,
+      fmt::join(_state_snapshots, ", "));
+
+    auto it = std::upper_bound(
+      _state_snapshots.begin(),
+      _state_snapshots.end(),
+      o,
+      [](model::offset o, const state_snapshot& current_update) {
+          return o < current_update.update_offset;
+      });
+
+    vassert(
+      it != _state_snapshots.begin(),
+      "Partition properties state machine holds an invariant that first "
+      "snapshot in state snapshots field is smaller than raft start offset "
+      "therefore the situation in which first element of the "
+      "_state_snapshots is returned is a critical failure. Requested "
+      "offset {}, state snapshots: {}",
+      o,
+      fmt::join(_state_snapshots, ", "));
+
+    // move iterator one step back to obtain cached snapshot with smaller
+    // offset
+    --it;
+    vlog(
+      _log.trace,
+      "returning indexed snapshot with offset: {}, requested offset: {}",
+      it->update_offset,
+      o);
+    co_return serde::to_iobuf(
+      raft_snapshot{.writes_disabled = it->writes_disabled});
+}
+
+ss::future<> partition_properties_stm::apply_local_snapshot(
+  raft::stm_snapshot_header header, iobuf&& buffer) {
+    vlog(_log.debug, "Applying local snapshot with offset {}", header.offset);
+    auto snapshot = serde::from_iobuf<local_snapshot>(std::move(buffer));
+    _state_snapshots = std::move(snapshot.state_updates);
+    co_return;
+}
+
+ss::future<raft::stm_snapshot> partition_properties_stm::take_local_snapshot(
+  ssx::semaphore_units apply_units) {
+    vlog(_log.debug, "Taking local snapshot");
+
+    auto last_applied = last_applied_offset();
+
+    // cleanup snapshot index
+    // delete all obsolete snapshot indicies
+    auto it = std::upper_bound(
+      _state_snapshots.begin(),
+      _state_snapshots.end(),
+      _raft->start_offset(),
+      [](model::offset o, const state_snapshot& current_update) {
+          return o < current_update.update_offset;
+      });
+    vassert(
+      it != _state_snapshots.begin(),
+      "State snapshot vector must always contain element that is strictly "
+      "smaller than the underlying log start offset. Snapshot vector {}",
+      fmt::join(_state_snapshots, ", "));
+    // decrease iterator to preseve entry smaller than the start offset
+    --it;
+    // cleanup only if there are elements to remove
+    if (it != _state_snapshots.begin()) {
+        chunked_vector<state_snapshot> state_updates;
+        state_updates.reserve(std::distance(it, _state_snapshots.end()));
+        // keep last snapshot with offset smaller than or equal to raft log
+        // start offset to be able to preserve the state from before the
+        // snapshot was taken
+        std::copy(
+          it, _state_snapshots.end(), std::back_inserter(state_updates));
+        _state_snapshots = std::move(state_updates);
+    }
+
+    auto state = _state_snapshots.copy();
```

Comment:
```suggestion
    chunked_vector<state_snapshot> state;
    if (it == _state_snapshots.begin()) {
        state = _state_snapshots.copy();
    } else {
        state.reserve(std::distance(it, _state_snapshots.end()));
        // keep last snapshot with offset smaller than or equal to raft log
        // start offset to be able to preserve the state from before the
        // snapshot was taken
        std::copy(
          it, _state_snapshots.end(), std::back_inserter(state));
        std::swap(_state_snapshots, state);
    }
```

Reply:
this suggestion has a bit different semantics as we wold snapshot a bit more than ti is actually needed, after the swap the `state` will contain all, not cleaned up entries